

---

**Overview**  
This series of scripts is designed for training and evaluating the DeepCaste model, a deep learning framework built on a CNN-RNN hybrid architecture. The model is trained using differentially accessible regions (500 bp) across social castes as input, categorized into four groups: nurse, forager, queen, and drone. For regions overlapping multiple castes, multiple labels were assigned, enabling the use of a multilabel and multiclass deep learning approach.

---

### **0_installation.sh**  
This script lists the required Python packages for building and training the DeepCaste model.

---

### **01_bed2fa.py**  
This script processes input regions in BED format, each with a fixed length of 500 bp, and converts them into FASTA sequences using the honeybee whole-genome sequence from NCBI. To expand the dataset, the regions were augmented by extending them to 700 bp around their summits and applying a sliding window of 500 bp with a 10-bp stride, increasing the sample size by 20-fold. The augmented dataset was then randomly split into training, testing, and cross-validation sets in an 8:1:1 ratio.  

**Input File Format**  
The input file consists of six columns:  
1. **Chromosome**: Name of the chromosome.  
2. **Start**: Starting position of the region on the chromosome.  
3. **End**: Ending position of the region on the chromosome.  
4. **Name**: Name of the region, formatted as `Chromosome:Start-End`.  
5. **Group**: The group where the region shows differential accessibility (Nurse (N), Forager (F), Queen (Q), Drone (D)).  
6. **Gene**: The gene closest to the region.  

**Example Input**:  
```plaintext
Chromosome	Start	End	Name	Group	Gene
NC-037638.1	89677	90177	NC-037638.1:89677-90177	N	LOC113219112
NC-037638.1	125901	126401	NC-037638.1:125901-126401	N	LOC113219398
NC-037638.1	1051852	1052352	NC-037638.1:1051852-1052352	N	LOC551678
NC-037638.1	1642523	1643023	NC-037638.1:1642523-1643023	N	LOC724948
```  

**Output Files**:  
- `train.fa`  
- `validation.fa`  
- `test.fa`  

---

### **02_train.py**  
This script trains the DeepCaste model using `train.fa`, `validation.fa`, and `test.fa` as inputs, generating model parameters saved as `model_best_loss.hdf5`. Custom functions required for training are organized in `utils.py`.  

**Model Architecture**:  
- **Conv1D layer**: 128 filters, kernel size of 20, stride of 1, ReLU activation.  
- **MaxPooling1D layer**: Pool size and stride of 10.  
- **Bidirectional LSTM layer**: 128 units with dropout (0.1) and recurrent dropout (0.1).  
- **Dense layer**: TimeDistributed Dense followed by a sigmoid activation with 4 output units for caste prediction.  
- **Dropout**: Applied after MaxPooling1D, Bidirectional LSTM, and Dense layers with fractions of 0.2, 0.2, and 0.4, respectively.  

Each region is one-hot encoded as a 500 bp Ã— 4-nucleotide sequence, with both forward and reverse strands passed through the model. Predictions average the neuron activations in the last Dense layer. The model is trained with the Adam optimizer (learning rate = 0.001) using binary cross-entropy loss (`binary_crossentropy`) for 32 epochs with a batch size of 128.

---

### **03_performance.py**  
This script evaluates model performance across training, validation, testing, and label-shuffled datasets. Performance metrics are calculated for each caste separately.  

**Metrics**:  
- Area under the Precision-Recall curve (auPR)  
- Area under the Receiver Operating Characteristic curve (auROC)  

These metrics are computed using the `average_precision_score` and `roc_auc_score` functions from the scikit-learn package.

---

### **04_DE_score.py**  
This script evaluates the importance of each nucleotide in the 500 bp input sequences.  

**Methodology**:  
1. **Saturation Mutagenesis**: Each nucleotide in the sequence is systematically substituted with A, C, G, or T, and changes in caste prediction scores are measured.  
2. **SHAP Analysis**: DeepExplainer from SHAP (v0.29.3) is used to compute nucleotide importance scores. The tool initializes with 500 random sequences and computes importance scores for each base position.

---

### **05_DE_plot.py**  
This script visualizes the nucleotide importance scores generated by SHAP.  


**Additional File Required**:  
- `high_light.txt`: Used to specify the regions that need to be highlighted.


**Example Input**:  
```plaintext
seq	TF	color	start	end
NC-037638.1:89677-90177	tgo	cyan	100	108
NC-037638.1:89677-90177	tgo	orange	146	153


**Visualizations**:  
- **Nucleotide Letter Plots**: Importance scores are multiplied by the one-hot encoded matrix of corresponding sequences and visualized as nucleotide letter heights.  
- **Dot Plots**: Variations in predicted scores during single-base knockout simulations are displayed as dot plots.

--- 
